File: 51-NQueens/51-NQueens.py
# Last updated: 6/11/2026, 11:37:47 AM
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        def is_safe(row, col, board):
            # Check for queens in the same column
            for i in range(row):
                if board[i][col] == 'Q':
                    return False
            
            # Check for queens in the upper left diagonal
            i, j = row - 1, col - 1
            while i >= 0 and j >= 0:
                if board[i][j] == 'Q':
                    return False
                i -= 1
                j -= 1
            
            # Check for queens in the upper right diagonal
            i, j = row - 1, col + 1
            while i >= 0 and j < n:
                if board[i][j] == 'Q':
                    return False
                i -= 1
                j += 1
            
            return True
        
        def backtrack(row, board, result):
            if row == n:
                result.append([''.join(row) for row in board])
                return
            
            for col in range(n):
                if is_safe(row, col, board):
                    board[row][col] = 'Q'
                    backtrack(row + 1, board, result)
                    board[row][col] = '.'
        
        board = [['.' for _ in range(n)] for _ in range(n)]
        result = []
        backtrack(0, board, result)
        return result

File: 51-NQueens/51-NQueens.cpp
// Last updated: 6/11/2026, 11:37:41 AM
class Solution {
public:

    vector<vector<string>> ans;
    bool isplacable(int r, int c, vector<string> &board){
        if(r<0 || c<0 || r>=board.size() || c>=board[0].size()){
            return false;
        }

        int dr = r;
        int dc = c;

        while(dr>=0 && dc>=0){
            if(board[dr][dc]!='Q'){
                dr--;
                dc--;
            }else{
                return false;
            }
        }

        dr = r;
        dc = c;
        while(dr>=0 && dc>=0){
            if(board[dr][dc]!='Q'){
                dr--;
            }else{
                return false;
            }
        }

        dr = r;
        dc = c;

        while(dr>=0 && dc<board.size()){
            if(board[dr][dc]!='Q'){
                dr--;
                dc++;
            }else{
                return false;
            }
        }

        return true;
    }


    void rec(int ind, vector<string> &board){
        if(ind == board.size()){
            ans.push_back(board);
            // return true;
            return;
        }
        int n = board.size();
        for(int i=0; i<n; i++){
            if(board[ind][i] == '.'){
                if(isplacable(ind, i, board)){
                    board[ind][i] = 'Q';
                    rec(ind+1, board);
                    board[ind][i] = '.';
                }
            }
        }

        return;
    }

    vector<vector<string>> solveNQueens(int n) {
        vector<string> board(n,string(n, '.'));
        ans.clear();
        rec(0, board);
        return ans;
    }
};